### ISSUE TITLE
--d4 Creates d4 files with inf/NaN values

### ISSUE THREAD

**Issue by cademirch** [brentp/mosdepth#237]
Been doing some work in the d4 repo and wanted to compare d4 files made using `d4tools create` and `mosdepth`. However with the small BAMs I've been testing with, `mosdepth` created d4 files have inf/NaN values for depth. Not sure if this is an issue with d4 or mosdepth.

**Mosdepth versions tested:** `0.3.6` (from conda) and `0.3.8` (the `mosdepth_d4` binary in the releases page)

d4tools was built from source in a fresh clone of the repo. 

**d4tools create**
```
> d4tools create -Azr ../data/genome.fna.fai ../data/test.bam test.d4
> d4tools show test.d4 > show.out
> head show.out
NC_052520.2	0	116790	0
NC_052520.2	116790	116942	1
NC_052520.2	116942	116946	0
NC_052520.2	116946	117093	1
NC_052520.2	117093	138700	0
NC_052520.2	138700	138848	1
NC_052520.2	138848	138852	2
NC_052520.2	138852	139000	1
NC_052520.2	139000	225154	0
NC_052520.2	225154	225306	1
```

**mosdepth 0.3.6**
```
> mosdepth --d4 test ../data/test.bam
> d4tools show test.per-base.d4 > show.out
> head show.out
NC_052520.2	0	36240	NaN
NC_052520.2	36240	36391	inf
NC_052520.2	36391	36426	NaN
NC_052520.2	36426	36577	inf
NC_052520.2	36577	43701	NaN
NC_052520.2	43701	43971	inf
NC_052520.2	43971	116790	NaN
NC_052520.2	116790	116941	inf
NC_052520.2	116941	116946	NaN
NC_052520.2	116946	117092	inf
```

0.3.8 is the same output as 0.3.6, no need to show it here.

Attached zipped folder of all the files for reference. 
[mosdepth-test.zip](https://github.com/user-attachments/files/16678568/mosdepth-test.zip)

**Comment by cademirch** [comment]
Did a little more testing today, and found that with mosdepth `0.3.3` I cannot reproduce this.

**Comment by brentp** [comment]
thanks for looking into this. I will have a look this month and see if it's easy to fix this issue. I'm not sure about supporting d4 output much longer.

**Comment by cademirch** [comment]
No problem. We do use mosdepth's d4 support so happy to help however. 

Out of curiosity, how come you're thinking of dropping d4?

**Comment by brentp** [comment]
well, I suppose the fact that you use it and that you and others are working on d4 is a good reason to keep it.

**Comment by brentp** [comment]
Hi, this is a problem, AFAICT with the bindings within d4 itself. the way that I test is this:
```
# in d4-format checkout
cargo build --release
cd d4binding/example/
make
./bin/create # this writes /tmp/test.d4
../../target/release/d4tools show /tmp/test.d4
```
where the final command shows this:
```
chr1	0	1	NaN
chr1	1	2	inf
chr1	2	3	inf
chr1	3	4	inf
chr1	4	5	inf
chr1	5	6	inf
chr1	6	7	inf
chr1	7	8	inf
chr1	8	9	inf
chr1	9	10	inf
chr1	10	11	inf
chr1	11	1001	NaN
chr1	1001	1002	inf
chr1	1002	1003	inf
chr1	1003	1004	inf
chr1	1004	1005	inf
chr1	1005	1006	inf
chr1	1006	1007	inf
chr1	1007	1008	inf
chr1	1008	1009	inf
chr1	1009	1010	inf
chr1	1010	1011	inf
chr1	1011	2000	NaN
chr1	2000	3000	inf
chr1	3000	4000	inf
chr1	4000	5000	inf
chr1	5000	10000	NaN
chr2	0	5001	NaN
chr2	5001	5002	inf
chr2	5002	5003	inf
chr2	5003	5004	inf
chr2	5004	5005	inf
chr2	5005	5006	inf
chr2	5006	5007	inf
chr2	5007	5008	inf
chr2	5008	5009	inf
chr2	5009	5010	inf
chr2	5010	5011	inf
chr2	5011	20000	NaN
```

**Comment by brentp** [comment]
I will look into this now. I would guess there is a type mismatch where it's getting i32 and expecting some other type.

**Comment by cademirch** [comment]
Thanks for tracking this down!

**Comment by brentp** [comment]
I have a fix.

**Comment by brentp** [comment]
fixed in latest release v0.3.9
thanks for reporting!
